File: tests/test_web_search.py
#!/usr/bin/env python3
"""Test Konoha's zero-API-key fallback search chain, caching, and fallback execution."""
import json
import os
import re
import urllib.request
import urllib.error
import urllib.parse
import unittest
from unittest.mock import patch, MagicMock

# Import search function directly from server module
import sys
sys.path.append(os.path.expanduser("~/.konoha"))
try:
    import server
except ImportError:
    # Fallback to local import if ~/.konoha not in path yet
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "src")))
    import server

def mock_urlopen_selector(url_or_req, timeout=None):
    if isinstance(url_or_req, urllib.request.Request):
        url = url_or_req.full_url
    else:
        url = url_or_req

    mock_resp = MagicMock()
    mock_resp.__enter__.return_value = mock_resp
    
    if "instances.json" in url:
        mock_resp.read.return_value = json.dumps({
            "instances": {
                "https://searx.test": {
                    "uptime": {"uptimeDay": 99.0},
                    "timing": {"search": {"all": {"median": 0.05}}}
                }
            }
        }).encode("utf-8")
    elif "searx.test/search" in url:
        if "q=test" in url:
            # Resolving test url during verification
            mock_resp.read.return_value = json.dumps({"results": []}).encode("utf-8")
        else:
            mock_resp.read.return_value = json.dumps({
                "results": [
                    {"title": "Test Svelte", "url": "https://svelte.dev", "content": "Svelte component framework"}
                ]
            }).encode("utf-8")
    elif "duckduckgo" in url:
        mock_resp.read.return_value = b"""
        <div class="result"><a class="result__a" href="/l/?kh=-1&amp;uddg=https://svelte.dev">Test Svelte</a><a class="result__snippet">Svelte component framework</a></div>
        """
    elif "startpage" in url:
        mock_resp.read.return_value = b"""
        <a class="result-link" href="https://svelte.dev"><h2 class="wgl-title">Test Svelte</h2></a>
        <p class="description">Svelte component framework</p>
        """
    elif "wikipedia" in url:
        mock_resp.read.return_value = json.dumps([
            "Python",
            ["Python (programming language)"],
            ["A high-level programming language."],
            ["https://en.wikipedia.org/wiki/Python_(programming_language)"]
        ]).encode("utf-8")
    else:
        raise urllib.error.URLError("Not mocked")
    return mock_resp

class TestWebSearchChain(unittest.TestCase):
    def test_query_simplification(self):
        """Verify that long queries simplify correctly by splitting words."""
        query = "how to implement sveltekit 3d threlte framework in production"
        terms = query.split()
        self.assertGreater(len(terms), 5)
        
        shortened = []
        for i in range(min(3, len(terms))):
            simp_q = " ".join(terms[:len(terms) - i])
            shortened.append(simp_q)
        
        self.assertEqual(len(shortened), 3)
        self.assertEqual(shortened[0], query)
        self.assertEqual(shortened[1], "how to implement sveltekit 3d threlte framework in")

    @patch("urllib.request.urlopen", side_effect=mock_urlopen_selector)
    def test_searxng_parser(self, mock_urlopen):
        """Verify SearXNG JSON parsing handles results and returns expected fields."""
        # Ensure we delete cache file so it runs dynamic check
        best_cache = os.path.expanduser("~/.konoha/searxng/best_instance.json")
        if os.path.exists(best_cache):
            try:
                os.remove(best_cache)
            except:
                pass

        res = server.run_web_search("Svelte", num_results=1)
        data = json.loads(res)
        
        self.assertEqual(data["status"], "success")
        self.assertGreater(len(data["results"]), 0)
        first_res = data["results"][0]
        self.assertEqual(first_res["title"], "Test Svelte")
        self.assertEqual(first_res["url"], "https://svelte.dev")
        self.assertEqual(first_res["snippet"], "Svelte component framework")

    @patch("urllib.request.urlopen")
    @patch("sys.stderr")
    def test_wikipedia_fallback(self, mock_stderr, mock_urlopen):
        """Verify Wikipedia OpenSearch API fallback is invoked when other engines fail."""
        # Force other engines to raise an error
        mock_urlopen.side_effect = urllib.error.URLError("Force Wikipedia Fallback")
        
        # We must allow the Wikipedia URL to succeed
        def mock_wiki_only(url_or_req, timeout=None):
            url = url_or_req.full_url if isinstance(url_or_req, urllib.request.Request) else url_or_req
            if "wikipedia" in url:
                mock_resp = MagicMock()
                mock_resp.__enter__.return_value = mock_resp
                mock_resp.read.return_value = json.dumps([
                    "Python",
                    ["Python (programming language)"],
                    ["A high-level programming language."],
                    ["https://en.wikipedia.org/wiki/Python_(programming_language)"]
                ]).encode("utf-8")
                return mock_resp
            raise urllib.error.URLError("Blocked for fallback test")

        mock_urlopen.side_effect = mock_wiki_only

        res_str = server.run_web_search("Python Programming Language", num_results=2)
        res = json.loads(res_str)
        self.assertEqual(res["status"], "success")
        self.assertGreaterEqual(res["results_count"], 0)
        if res["results"]:
            self.assertEqual(res["results"][0]["source"], "Wikipedia")

if __name__ == "__main__":
    unittest.main()

File: tests/test_web_search.js
#!/usr/bin/env node
'use strict';

/**
 * tests/test_web_search.js — Test Konoha's zero-API-key fallback search chain, caching, and fallback execution.
 */

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const server = require('../src/server');

async function run() {
  console.log('Running test_web_search tests...');
  const originalFetch = global.fetch;
  const originalSearxUrl = process.env.SEARXNG_URL;

  try {
    // 1. Query simplification
    const query = 'how to implement sveltekit 3d threlte framework in production';
    const terms = query.split(/\s+/);
    assert.ok(terms.length > 5);

    const shortened = [];
    for (let i = 0; i < Math.min(3, terms.length); i++) {
      const simpQ = terms.slice(0, terms.length - i).join(' ');
      shortened.push(simpQ);
    }
    assert.strictEqual(shortened.length, 3);
    assert.strictEqual(shortened[0], query);
    assert.strictEqual(shortened[1], 'how to implement sveltekit 3d threlte framework in');
    console.log('✓ Query simplification passed');

    // 2. SearXNG parser and null safety
    const bestCache = path.join(os.homedir(), '.konoha', 'searxng', 'best_instance.json');
    const instCache = path.join(os.homedir(), '.konoha', 'searxng', 'instances_cache.json');
    for (const c of [bestCache, instCache]) {
      try { if (fs.existsSync(c)) fs.unlinkSync(c); } catch (_) {}
    }

    global.fetch = async (url, opts) => {
      const urlStr = String(url);
      if (urlStr.includes('instances.json')) {
        return {
          ok: true,
          json: async () => ({
            instances: {
              'https://searx.test': {
                uptime: { uptimeDay: 99.0 },
                timing: { search: { all: { median: 0.05 } } }
              },
              'https://searx.broken-null-uptime': {
                uptime: null,
                timing: null
              },
              'https://searx.broken-null-search': {
                uptime: { uptimeDay: 98.0 },
                timing: { search: null }
              }
            }
          })
        };
      } else if (urlStr.includes('searx.test/search') || urlStr.includes('searx.test')) {
        if (urlStr.includes('q=test')) {
          return {
            ok: true,
            json: async () => ({ results: [] })
          };
        }
        return {
          ok: true,
          json: async () => ({
            results: [
              { title: 'Test Svelte', url: 'https://svelte.dev', content: 'Svelte component framework' }
            ]
          })
        };
      }
      throw new Error('Not mocked: ' + urlStr);
    };

    const res1Str = await server.runWebSearch('Svelte', 1);
    const data1 = JSON.parse(res1Str);
    assert.strictEqual(data1.status, 'success');
    assert.ok(data1.results.length > 0);
    assert.strictEqual(data1.results[0].title, 'Test Svelte');
    assert.strictEqual(data1.results[0].url, 'https://svelte.dev');
    assert.strictEqual(data1.results[0].snippet, 'Svelte component framework');
    console.log('✓ SearXNG parser and null safety passed');

    // 3. SearXNG env override
    process.env.SEARXNG_URL = 'https://custom-searx.internal';
    global.fetch = async (url, opts) => {
      const urlStr = String(url);
      if (urlStr.includes('custom-searx.internal')) {
        return {
          ok: true,
          json: async () => ({
            results: [
              { title: 'Test Svelte', url: 'https://svelte.dev', content: 'Svelte component framework' }
            ]
          })
        };
      }
      throw new Error('Not mocked: ' + urlStr);
    };

    const res2Str = await server.runWebSearch('Svelte', 1);
    const data2 = JSON.parse(res2Str);
    assert.strictEqual(data2.status, 'success');
    assert.ok(data2.results.length > 0);
    assert.ok(data2.results[0].source.includes('custom-searx.internal'));
    console.log('✓ SearXNG env override passed');

    // 4. Wikipedia fallback
    delete process.env.SEARXNG_URL;
    for (const c of [bestCache, instCache]) {
      try { if (fs.existsSync(c)) fs.unlinkSync(c); } catch (_) {}
    }

    global.fetch = async (url, opts) => {
      const urlStr = String(url);
      if (urlStr.includes('wikipedia')) {
        return {
          ok: true,
          json: async () => [
            'Python',
            ['Python (programming language)'],
            ['A high-level programming language.'],
            ['https://en.wikipedia.org/wiki/Python_(programming_language)']
          ]
        };
      }
      throw new Error('Engine unavailable');
    };

    const res3Str = await server.runWebSearch('Python Programming Language', 2);
    const data3 = JSON.parse(res3Str);
    assert.strictEqual(data3.status, 'success');
    assert.ok(data3.results_count >= 1);
    if (data3.results && data3.results.length > 0) {
      assert.strictEqual(data3.results[0].source, 'Wikipedia');
    }
    console.log('✓ Wikipedia fallback passed');

    console.log('\nAll test_web_search tests passed!');
  } finally {
    global.fetch = originalFetch;
    if (originalSearxUrl !== undefined) {
      process.env.SEARXNG_URL = originalSearxUrl;
    } else {
      delete process.env.SEARXNG_URL;
    }
  }
}

if (require.main === module) {
  run().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { run };
